Allow viewing of submitted answers

dwyeradam opened the pull request:
### What is the context of this PR?
**As a** respondent, 
**I need** to access my responses to questions 
**so that** I can refer to them if my data is queried by ONS at a later date or used as a reference for next time I complete the same survey

### How to review 
Easiest to test using docker (`docker-compose up -d`) to ensure the dynamoDb container is created.

Use `test_submitted_responses.json` schema which allows answers to be viewed for 45 seconds.

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,83 @@
+
+    def __init__(self, table_name='submitted-responses'):
+        if EQ_ENVIRONMENT_PREFIX == 'dev':
+            self.dynamoDB = boto3.resource('dynamodb', endpoint_url='http://localhost:6060')
```
I think we should explicitly pass in the endpoint and call this line only if one is passed in rather then switching on a hard coded value of 'dev'

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,83 @@
+    def insert_item(self, tx_id, answers, metadata, submitted, time_to_live=EQ_SUBMISSION_VALIDITY):
+        """Insert an item to table"""
+
+        if not time_to_live or time_to_live > EQ_SUBMISSION_VALIDITY:
```
Can `time_to_live` be none of we default it in the method?

dwyeradam replied:
Removed default.

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,83 @@
+                raise ClientError('Could not load data into DynamoDB table', 'put_item')
+        except ClientError:
+            logger.bind(tx_id=tx_id)
+            logger.info('submitted_responses dynamoDB', 'Could not load data into DynamoDB table:')
```
This log line contains a trailing `:`

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,83 @@
+        }
+
+        try:
+            table = self.dynamoDB.Table(self.tableName)
```
you could set `self.table` in the constructor instead of `self.tableName`

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,83 @@
+            response = table.get_item(Key={'tx_id': tx_id})
+            return response['Item']
+        except ClientError:
+            return None
```
If the response is empty we should return None. If we get an exception we should log it.

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,83 @@
+            )
+            return response['Item']['valid_until']
+        except (KeyError, ClientError):
+            return 0
```
same as above

jonnyshaw89 commented on app/libs/utils.py:
```diff
@@ -2,6 +2,11 @@
+    return (tx_id[:4] + '-' + tx_id[4:])[:19]
+
+
+def convert_tx_id_for_boxes(tx_id):
```
Do we use the boxes anymore? if not we should remove this.

dwyeradam replied:
Used in an error page but i'll need to check what context that error page would get used.

jonnyshaw89 replied:
It might been good to make the error page show the tx_id in a consistent way to the thank you?

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -260,7 +312,7 @@ def _is_skipping_to_the_end(routing_path, block, current_location):
 
 
-def submit_answers(routing_path, eq_id, form_type, collection_id):
+def submit_answers(routing_path, eq_id, form_type, collection_id, block_id):
```
rather than pass the `block_id` into this method, we have the `g.schema_json` so just lookup if it has a summary

dwyeradam replied:
Going to refactor the way we define if we show the submitted responses or not. This will come out as part of that. In future we'll define it within an object in the schema so it won't be dependent on having a summary page.

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -538,8 +603,11 @@ def _get_block_json(current_location):
-def _get_schema_context(full_routing_path, current_location, metadata, answer_store):
-    aliases = SchemaHelper.get_aliases(g.schema_json)
+def _get_schema_context(full_routing_path, current_location, metadata, answer_store, schema=None):
+    if schema:
```
Im not sure why we have done this?

dwyeradam replied:
You're right, it's no longer required. It was left in from a time when the schema wasn't available after submission. That however was fixed.

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,81 @@
+class SubmittedResponses(object):
+
+    def __init__(self):
+        if EQ_DYNAMODB_ENDPOINT:
```
if `EQ_DYNAMODB_ENDPOINT` defaults to None, what happens if you pass None into the `endpoint_url`?
ie, do you need the if/else?

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,81 @@
+            else:
+                raise BotoConnectionError()
+        except (BotoCoreError, BotoConnectionError):
+            logger.bind(tx_id=tx_id)
```
Is `tx_id` not already bound? if not we should do it towards the beginning of the request

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,81 @@
+        try:
+            response = self.table.get_item(Key={'tx_id': tx_id})
+            return response['Item']
+        except (BotoCoreError, BotoConnectionError, KeyError):
```
We should also log these exceptions?

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,81 @@
+        """
+        try:
+            response = self.table.get_item(Key={'tx_id': tx_id})
+            return response['Item']
```
If there are no results does the 'Item' key exist?

dwyeradam replied:
I've badly used Exception handling to deal with expected behaviour which I'll refactor.

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -172,22 +181,59 @@ def post_household_composition(routing_path, **kwargs):
 
     if survey_completed_metadata:
+        view_submission = None
+        tx_id = survey_completed_metadata['tx_id']
```
Put this inside the If, as we dont need it outside

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -273,6 +319,7 @@ def submit_answers(routing_path, eq_id, form_type, collection_id):
             routing_path,
         )
 
+        message = json.dumps(message)
```
why are we doing this? does it change the structure of what we encrypt?

dwyeradam replied:
See #1329

shahi645 replied:
@dwyeradam this has been changed to be done earlier : `message = json.dumps(convert_answers(`

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -300,11 +360,19 @@ def submit_answers(routing_path, eq_id, form_type, collection_id):
 
+def store_submitted_responses():
+    if g.schema_json.get('view_submitted_responses'):
+        if g.schema_json['view_submitted_responses'].get('enabled'):
```
what if enabled is set to `False`?

jonnyshaw89 replied:
Assuming it returns a boolean then you can just return this line? rather than it being an if statement.

jonnyshaw89 commented on docker-compose.yml:
```diff
@@ -44,6 +44,8 @@ services:
+      EQ_ENVIRONMENT_PREFIX: dev
+      EQ_DYNAMODB_ENDPOINT: http://localhost:6060
     restart: always
     depends_on:
```
Add a depends on for `eq-docker-dynamodb`

jonnyshaw89 commented on docker-compose.yml:
```diff
@@ -44,6 +44,8 @@ services:
       EQ_SECRETS_FILE: docker-secrets.yml
       EQ_KEYS_FILE: docker-keys.yml
+      EQ_ENVIRONMENT_PREFIX: dev
+      EQ_DYNAMODB_ENDPOINT: http://localhost:6060
```
you should route the communication within the docker network by changing this to `http://eq-docker-dynamodb:6060`

ajmaddaford commented:
- The answers need to be stored encrypted exactly as they are when being stored in Postgres. For this we will need to keep the user signed in (so we can encrypt and decrypt the answers with the user provided keys). 
- We should look at doing this by adding configurable backends for questionnaire store. This could potentially allow us to switch out Postgres for DynamoDB for survey state during surveys.
- The extra information required on the thank you and answers after submission screen e.g. ru name, submission time, should be stored in the server side session. 
- In doing this we should also be able to store the eq id and form type in the session and remove them from the urls for post submission screens.

shahi645 commented:
@dwyeradam still getting 5xx's on submission, to do with Boto3's `PutItem` class

jonnyshaw89 commented:
@dwyeradam @shahi645 this is because there is an extra `$` on line 36 of `dev_settings.sh`

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -276,13 +333,32 @@ def submit_answers(routing_path, eq_id, form_type, collection_id):
+                }
+            )
+
+            SubmittedResponses().insert_item(metadata['tx_id'], encrypted_data, submitted_time, submission_validity)
```
I think we should look at if we can extend `EncryptedQuestionnaireStorage` and provide multiple implementations then we can just update the storage mechanism on `QuestionnaireStore` and save a copy?
This would also set us in a good place for if we want to use DynamoDB for storing answers eventually

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -293,11 +369,18 @@ def submit_answers(routing_path, eq_id, form_type, collection_id):
         return redirect(invalid_location.url(metadata))
 
 
+def store_submitted_responses():
```
I think this should be renamed as it reads as if the function is doing the storing rather then returning a boolean

dwyeradam replied:
Changing to is_submission_stored

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -293,11 +369,18 @@ def submit_answers(routing_path, eq_id, form_type, collection_id):
 
     for location in full_routing_path:
-        if location.block_id in ['thank-you', 'summary', 'confirmation']:
+        if location.block_id in ['thank-you', 'summary', 'confirmation', 'view-submission']:
```
I think these need to be changed to look at the Block type. But that may be another PR?

hamishtaplin commented on app/templates/thank-you.html:
```diff
@@ -9,19 +9,37 @@ <h1 class="saturn">Submission Successful</h1>
-    <p class="u-mb-no mars">You may wish to save or print this page for your records.</p>
+<div class="u-mb-s">
+    <div class="panel panel--simple panel--success panel--spacious">
+        <p class=""mars>Your answers were submitted for <b>{{meta.respondent.ru_name}}</b> on {{meta.survey.submitted_time}}</p>
```
The class name is outside the quotes.

hamishtaplin commented on app/templates/thank-you.html:
```diff
@@ -9,19 +9,37 @@ <h1 class="saturn">Submission Successful</h1>
+{% if meta.survey.submission_validity > 0 %}
+    <div class="u-mb-s">
+    {% if view_submission_url %}
+        <a class="btn btn--primary btn--lg" data-qa="btn-view-submission" href={{view_submission_url}}>View and print a copy of your answers</a>
```
Links shouldn't look like buttons.

dwyeradam replied:
@hamishtaplin, @armstrongb If we don't use a button I need some advice on how to make this link a primary feature of the page please.

hamishtaplin commented on app/templates/view-submission.html:
```diff
@@ -0,0 +1,73 @@
+
+<div class="u-mb-s">
+    <div class="panel panel--simple panel--success panel--spacious">
+        <p class=""mars>Your answers were submitted for <b>{{meta.respondent.ru_name}}</b> on {{meta.survey.submitted_time}}</p>
```
Class outside the quotes.

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,64 @@
+    def insert_item(self, tx_id, data, submitted, time_to_live):
+        """Insert an item to table"""
+
+        if not time_to_live or time_to_live > EQ_SUBMISSION_VALIDITY:
```
i don't think we need `if not time_to_live` as `time_to_live` is always set

jonnyshaw89 commented on app/settings.py:
```diff
@@ -73,6 +76,7 @@ def get_env_or_fail(key):
 
 EQ_JWT_LEEWAY_IN_SECONDS = 120
 DEFAULT_LOCALE = 'en_GB'
+EQ_SUBMISSION_VALIDITY = 900
```
Was this now mandatory in the schema?

jonnyshaw89 commented on app/storage/storage_encryption.py:
```diff
@@ -0,0 +1,72 @@
+        sha256.update(to_str(pepper).encode('utf-8'))
+
+        # we only need the first 32 characters for the CEK
+        return to_bytes(sha256.hexdigest()[:32])
```
if we wrap this in `generate_jwk_from_cek()` then it would save having to call `key = self.generate_jwk_from_cek(self._cek)` whenever we use it.

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -199,6 +215,50 @@ def get_thank_you():
+
+            metadata_context = build_metadata_context_for_survey_completed(vars(session_data))
+
+            current_location = Location('group', 0, 'view-submission')
```
This code smells a bit. Why do we have a hard-coded Location?

jonnyshaw89 replied:
The only thing that this location is used for is to go into `_get_schema_context()` and that only uses the `group_instance` so it might be worth refactoring `_get_schema_context()` to just take in the `group_instance`

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -199,6 +215,50 @@ def get_thank_you():
+                                                             summary=summary_rendered_context)
+            return submission_view_template
+
+    return get_thank_you()
```
I think we should re-direct to the thank you page otherwise you will have the thank you page with the url of `/view-submission`

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -283,17 +343,47 @@ def submit_answers(routing_path):
+        session_data = session_store.session_data
+        session_data.submitted_time = submitted_time.isoformat()
+        session_data.submission_validity = submission_validity
+        session_store.session_data = session_data
```
Is this line needed? we are editing the object so no need to set it back?

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -283,17 +343,47 @@ def submit_answers(routing_path):
+    if g.schema_json.get('view_submitted_responses'):
+        enabled = g.schema_json['view_submitted_responses'].get('enabled', False)
+        if enabled:
+            validity = g.schema_json['view_submitted_responses'].get('timeout', EQ_SUBMISSION_VALIDITY)
```
Didn't we stop this being optional in the schema?

shahi645 commented on app/data_model/answer_store.py:
```diff
@@ -122,6 +122,20 @@ def add_or_update(self, answer):
 
+    def add_multiple_answers(self, answers):
+        """
+        Add a new answers into the answer store, or update if it exists.
```
typo get rid of s

jonnyshaw89 replied:
you mean the `a` yeh?

shahi645 replied:
yes srry method is multiple answers, my bad

jonnyshaw89 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,61 @@
+logger = get_logger()
+
+
+class SubmittedResponses(object):
```
I think we should create this once for the application maybe in `setup.py` and then reuse it?

dwyeradam replied:
Done.

shahi645 commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,61 @@
+class SubmittedResponses(object):
+
+    def __init__(self):
+        config = Config(retries={'max_attempts': 1})   # Number of additional connection attempts
```
Should `max_attempts` be ore than 1 ?

dwyeradam replied:
We'll check the database is available twice as this is additional checks. It would be very unlikely it wouldn't be able to access it first time as it's AWS to AWS. Any further attempts would cause a time delay felt by the user.

shahi645 commented:
Looks good so far, just tests and print button to go! 👍

jonnyshaw89 commented on tests/functional/spec/features/submitted_responses.spec.js:
```diff
@@ -0,0 +1,64 @@
+          .pause(5000)
+          .click(ThankYouPage.viewSubmitted())
+          .getUrl().should.eventually.contain(ThankYouPage.pageName)
+          .getSource().should.eventually.contain('You are no longer able to view and print your answers.');
```
Can you access this text with a selector? it will make the error message better if so

jonnyshaw89 commented on tests/functional/spec/features/submitted_responses.spec.js:
```diff
@@ -0,0 +1,64 @@
+          .isVisible(ThankYouPage.viewSubmitted())
+          .pause(5000)
+          .refresh()
+          .getUrl().should.eventually.contain(ThankYouPage.pageName)
```
I don't think we need this check as the url shouldn't change on a refresh?

jonnyshaw89 commented on tests/functional/spec/features/submitted_responses.spec.js:
```diff
@@ -0,0 +1,64 @@
+  });
+
+    describe('Try to view submission after timeout', function() {
+    it('Given I complete to Thank-You page, When I refresh after the timeout period, Then I should not be able to view my submitted answers', function() {
```
I think the indentation is incorrect here.

jonnyshaw89 commented on app/storage/encrypted_questionnaire_storage.py:
```diff
@@ -42,14 +34,15 @@ def add_or_update(self, data, version):
-            if 'data' in data:
-                decrypted_data = self._decrypt_data(data['data'])
-                return decrypted_data, version
+        if data and 'data' in data:
```
If we keep this nested inside the `if` above then we don't need the extra code here?

ajmaddaford commented on app/templates/view-submission.html:
```diff
@@ -0,0 +1,73 @@
+</div>
+
+<div>
+  <div class="summary">
```
We should try and share this summary html with the summary screen - is the only difference no 'Edit' links?

ajmaddaford commented on app/templating/metadata_context.py:
```diff
@@ -51,8 +54,12 @@ def build_metadata_context_for_survey_completed(survey_completed_metadata):
     return {
         'survey': {
             'period_str': survey_completed_metadata['period_str'],
+            'submission_validity': round(Decimal(survey_completed_metadata['submission_validity'] / 60), 2),
```
`survey_completed_metadata` has been refactored into SessionData, can we pass the SessionData class into this method instead?

ajmaddaford commented on app/templating/metadata_context.py:
```diff
@@ -51,8 +54,12 @@ def build_metadata_context_for_survey_completed(survey_completed_metadata):
         'survey': {
             'period_str': survey_completed_metadata['period_str'],
+            'submission_validity': round(Decimal(survey_completed_metadata['submission_validity'] / 60), 2),
+            'submitted_time': dateutil.parser.parse(survey_completed_metadata['submitted_time']).strftime('%d %B %Y at %H:%M')
```
An ISO format date should be set here and the template should convert it (probably via a jinja filter) into the `%d %B %Y at %H:%M` format

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -224,6 +235,48 @@ def get_thank_you(eq_id, form_type):  # pylint: disable=unused-argument
+
+            answer_store = AnswerStore()
+            submitted_data = json.loads(encrypter.decrypt_data(submitted_data['data']))
+            answer_store.add_multiple_answers(submitted_data.get('answers'))
```
You can pass a dict of answers to AnswerStore on initialisation - `answer_store = AnswerStore(existing_answers= submitted_data.get('answers'))` - which should remove the need for the new `add_multiple_answers` method.

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -224,6 +235,48 @@ def get_thank_you(eq_id, form_type):  # pylint: disable=unused-argument
+def get_view_submission(eq_id, form_type):  # pylint: disable=unused-argument
+
+    session_data = get_session_store().session_data
+    submission_valid_until = int(dateutil.parser.parse(session_data.submitted_time).strftime('%s')) + session_data.submission_validity
```
I think it would be clearer if we used `datetime.timedelta` and compared date objects rather than converting dates to ints. Something like:
```
submission_time = dateutil.parser.parse(session_data.submitted_time)
submission_valid_until = submission_time + datetime.timedelta(seconds=submission_validity)
```
The comparison would then be:
```
if submission_valid_until > datetime.utcnow():
```

This applies to the other places where dates are converted to ints

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -303,20 +356,52 @@ def submit_answers(routing_path, eq_id, form_type):
             raise SubmissionFailedException()
 
-        get_questionnaire_store(current_user.user_id, current_user.user_ik).delete()
+        store_submitted_responses_and_session_data(answer_store.answers, metadata)
```
I don't think we should be combining this into one method. We always store the session data, but only store the submitted responses if it's enabled for the questionnaire.

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -303,20 +356,52 @@ def submit_answers(routing_path, eq_id, form_type):
+        if enabled:
+            validity = view_submitted_responses['timeout']
+
+    return enabled, validity
```
I think this method should just return whether the schema has view_submitted_responses enabled or not. We can enforce in the schema that timeout should be greater than 0.

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -303,20 +356,52 @@ def submit_answers(routing_path, eq_id, form_type):
+    session_store = get_session_store()
+    session_data = session_store.session_data
+    session_data.submitted_time = submitted_time.isoformat()
+    session_data.submission_validity = submission_validity
```
We should keep this simple and just store a boolean to say whether a submission has been persisted to DyanmoDB or not. When we need to later on we can get the timeout from the schema.

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -224,6 +237,45 @@ def get_thank_you(eq_id, form_type):  # pylint: disable=unused-argument
+            rendered_schema = renderer.render(g.schema.json, **schema_context)
+            summary_rendered_context = build_summary_rendering_context(g.schema, rendered_schema, answer_store, metadata)
+
+            submission_view_template = render_theme_template(g.schema.json['theme'],
```
You should just return here rather than assign a variable

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -208,13 +212,22 @@ def get_thank_you(eq_id, form_type):  # pylint: disable=unused-argument
+        metadata_context = build_metadata_context_for_survey_completed(session_data)
+
+        view_submission_url, view_submission_duration = None, 0
+        if session_data.is_viewable_submission_stored:
```
I think we should drop `is_viewable_submission_stored` and rely on checking the submitted time against the view submission duration. If nothing has gone wrong then it would work in the same way. If it hasn't been stored then we can error or redirect back to the thank-you page.

ajmaddaford commented on data/en/test_submitted_responses.json:
```diff
@@ -0,0 +1,113 @@
+    "view_submitted_responses": {
+        "enabled": true,
+        "timeout": 5
+    },
```
I think we can improve the naming here. It would be clearer if we renamed:
- `view_submitted_responses` to `view_submitted_response` as it's for a single questionnaire response, and
- `timeout` to `duration`

ajmaddaford reviewed (approved):
I think we need to have an environment variable to allow not using DynamoDB, similar to EQ_RABBITMQ_ENABLED. We'd have to make the EQ_SUBMITTED_RESPONSES_TABLE_NAME environment variable mandatory only when DynamoDB is enabled. This will also allow it to be disabled for the majority of the create_app tests that mock boto3.

ajmaddaford reviewed (changes requested):
I think we need to have an environment variable to allow not using DynamoDB, similar to EQ_RABBITMQ_ENABLED. We'd have to make the EQ_SUBMITTED_RESPONSES_TABLE_NAME environment variable mandatory only when DynamoDB is enabled. This will also allow it to be disabled for the majority of the create_app tests that mock boto3.

ajmaddaford reviewed (changes requested):
I think we need to have an environment variable to allow not using DynamoDB, similar to EQ_RABBITMQ_ENABLED. We'd have to make the EQ_SUBMITTED_RESPONSES_TABLE_NAME environment variable mandatory only when DynamoDB is enabled. This will also allow it to be disabled for the majority of the create_app tests that mock boto3.

jonnyshaw89 commented on tests/integration/integration_test_case.py:
```diff
@@ -40,12 +40,25 @@ def setUp(self):
+        # Perform setup steps
+        self._setUpApp(True)
+
+    def _setUpApp(self, enableDynamoDB=False):
```
variable names should have `_` rather then capitals.

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -224,6 +236,44 @@ def get_thank_you(eq_id, form_type):  # pylint: disable=unused-argument
+    g.schema = load_schema_from_params(session_data.eq_id, session_data.form_type)
+
+    if _is_submission_viewable(g.schema.json, session_data.submitted_time):
+        submitted_data = current_app.eq['submitted_responses'].get_item(session_data.tx_id)
```
we need to check here is `submitted_responses` in present
`if 'submitted_responses' in current_app.eq:`

dwyeradam replied:
Added check to `def is_view_submitted_response_enabled(schema):` so that we don't attempt anything around submitted answers unless avaliable.

ajmaddaford commented on app/data_model/session_store.py:
```diff
@@ -24,7 +26,7 @@ def create(self, eq_session_id, user_id, session_data):
         self.user_id = user_id
         self.session_data = session_data
-        self._eq_session = EQSession(self.eq_session_id, self.user_id, json.dumps(vars(self.session_data)))
+        self._eq_session = EQSession(self.eq_session_id, self.user_id, vars(self.session_data))
```
I think it would be clearer to not set `session_data` on `EQSession` until `save` is called, as this looks like we're storing session data unencrypted, which isn't the case.

ajmaddaford commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,44 @@
+    def __init__(self, table):
+        self.table = table
+
+    def insert_item(self, tx_id, data, submitted_time, time_to_live):
```
I think it would be clearer to call this method `put_item` to be consistent with the DynamoDB naming. I also highlights that `put` will replace an item if it already exists (not that this should happen).

ajmaddaford commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,44 @@
+            Item = response.get('Item', None)
+            if Item:
+                Item['valid_until'] = datetime.utcfromtimestamp(Item['valid_until'])
+            return Item
```
The variable `item` should be lowercase

ajmaddaford commented on app/data_model/submitted_responses.py:
```diff
@@ -0,0 +1,44 @@
+        """Insert an item to table"""
+
+        submitted_time = int(submitted_time.strftime('%s'))  # convert to epoch
+        valid_until = submitted_time + time_to_live
```
I think we should create `item` outside of the put method, so the get and put are consistent (the get returns an item dict). It could take a dict with valid until as a time and convert it to epoch in this method, which would also mirror what's done in the get method.

ajmaddaford commented on app/jinja_filters.py:
```diff
@@ -76,6 +76,12 @@ def format_month_year_date(value, date_format='%B %Y'):
 
 
+@blueprint.app_template_filter()
+def format_iso_date_string(value, date_format='%d %B %Y at %H:%M'):
```
- We should call this method `format_datetime` to be consistent with `format_date` and `format_month_year_date`
- We should drop the test variable and just return the string as done in the `format_date` and `format_month_year_date` methods

ajmaddaford commented on app/settings.py:
```diff
@@ -44,6 +44,7 @@ def get_env_or_fail(key):
 EQ_SECRETS_FILE = os.getenv('EQ_SECRETS_FILE', 'secrets.yml')
 EQ_KEYS_FILE = os.getenv('EQ_KEYS_FILE', 'keys.yml')
 
+EQ_SUBMITTED_RESPONSES_TABLE_NAME = get_env_or_fail('EQ_SUBMITTED_RESPONSES_TABLE_NAME')
```
This should only be mandatory if `EQ_DYNAMODB_ENABLED` is `True`

ajmaddaford commented on app/templates/thank-you.html:
```diff
@@ -9,19 +9,37 @@ <h1 class="saturn">Submission Successful</h1>
-    </ul>
-    {% endif %}
-    <p class="u-mb-no mars">You may wish to save or print this page for your records.</p>
+<div class="u-mb-s">
```
There is extra content on this screen that isn't in the designs. It looks like the "Thank you for submitting the <survey name>, your responses have been successfully received." and "It is now safe to close this window." has been dropped. We should check with @armstrongb.

dwyeradam replied:
Talked with @armstrongb and have removed both of these.

ajmaddaford commented on app/templates/thank-you.html:
```diff
@@ -9,19 +9,37 @@ <h1 class="saturn">Submission Successful</h1>
+        <div class="panel panel--simple panel--info" data-qa="view-submission-expired">
+            <div class="panel__body">
+                <p>You are no longer able to view and print your answers.</p>
+                <p>The link expires to maintain the security of your information.</p>
```
Where does this line come from? I can't see it in the designs.

dwyeradam replied:
This was agreed with @charlotte-hirst and the line is taken from the error page *ONS-screenshot-expired-print-link-error.png* that we decided not to use but felt this line gave additional useful information here.

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -208,14 +212,22 @@ def get_thank_you(eq_id, form_type):  # pylint: disable=unused-argument
+        view_submission_url, view_submission_duration = None, 0
+        if is_view_submitted_response_enabled(schema.json):
+            view_submission_duration = schema.json['view_submitted_response']['duration']
+            if _is_submission_viewable(schema.json, session_data.submitted_time):
```
Duration is only needed if the submission is viewable, so it would be better to move it to the point we work out the url. The logic for converting it to minutes should also move here.

dwyeradam replied:
I was using duration as the check for showing either the link or the timeout panel on the thank you page. If it was greater than 0 then it showed the view submission content. Instead i'm now passing a boolean of `is_view_submission_enabled`

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -208,14 +212,22 @@ def get_thank_you(eq_id, form_type):  # pylint: disable=unused-argument
-        return thank_you_template
+        metadata_context = build_metadata_context_for_survey_completed(session_data)
+
+        view_submission_url, view_submission_duration = None, 0
```
I don't think we gain anything by these being on one line and I think it would be clearer on two.

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -224,6 +236,44 @@ def get_thank_you(eq_id, form_type):  # pylint: disable=unused-argument
+    if _is_submission_viewable(g.schema.json, session_data.submitted_time):
+        submitted_data = current_app.eq['submitted_responses'].get_item(session_data.tx_id)
+
+        if submitted_data and submitted_data['valid_until'] > datetime.utcnow():
```
Do we need to check `submitted_data['valid_until'] > datetime.utcnow()` as it's already checked in the call to `_is_submission_viewable`?

dwyeradam replied:
No, probably a bit of overkill checking that both times match. I'll remove.

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -303,20 +353,59 @@ def submit_answers(routing_path, eq_id, form_type):
         return redirect(invalid_location.url(metadata))
 
 
+def _store_session_data(submitted_time):
```
It would be clearer if we named this method `_store_submitted_time_in_session`

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -303,20 +353,59 @@ def submit_answers(routing_path, eq_id, form_type):
+
+
+def _store_viewable_submission(answers, metadata, submitted_time):
+    if is_view_submitted_response_enabled(g.schema.json):
```
I think it would be clearer if the if statement was move outside of this method, to make it clear we only store the viewable submission if the schema enables it.

ajmaddaford commented on app/data_model/session_store.py:
```diff
@@ -69,8 +72,10 @@ def _load(self):
+
             if self._eq_session.session_data:
-                self.session_data = json.loads(self._eq_session.session_data, object_hook=lambda d: SessionData(**d))
+                session_data = StorageEncryption(self.user_id, self.user_ik).decrypt_data(self._eq_session.session_data)
```
This assumes that the session is encrypted, which isn't the case if someone already has an active session when we deploy this. We should be able to make it work with un-encrypted sessions on load, which will then be encrypted on save.

jonnyshaw89 commented on app/data_model/session_store.py:
```diff
@@ -74,7 +75,10 @@ def _load(self):
             if self._eq_session.session_data:
-                session_data = StorageEncryption(self.user_id, self.user_ik).decrypt_data(self._eq_session.session_data)
+                session_data = self._eq_session.session_data
+                if self._is_session_data_encrypted(session_data):
```
We should add a blocked card in the backlog to remove this.

jonnyshaw89 commented on app/settings.py:
```diff
@@ -67,7 +66,10 @@ def get_env_or_fail(key):
 
 EQ_DYNAMODB_ENABLED = parse_mode(os.getenv('EQ_DYNAMODB_ENABLED', 'True'))
-EQ_DYNAMODB_ENDPOINT = os.getenv('EQ_DYNAMODB_ENDPOINT', None)
+EQ_DYNAMODB_ENDPOINT = os.getenv('EQ_DYNAMODB_ENDPOINT')
```
Can the endpoint go inside the `if` too?

yoganandkunche reviewed (changes requested):
@dwyeradam I don't think we need the help us improve on the view answer page as the session is already expired it will only give an error 403.

hamishtaplin commented on app/templates/summary.html:
```diff
@@ -20,65 +18,22 @@
-  {% endfor %}
+{% if content.is_view_submission_enabled %}
+  <div class="u-mb-s u-mb-m@s">
+    <div class="venus download__text">You will have the opportunity to view and print a copy of your answers after submitting this questionnaire</div>
```
`download__text` class shouldn't be there

ajmaddaford replied:
"a copy of" seems redundant

hamishtaplin commented on app/templates/view-submission.html:
```diff
@@ -0,0 +1,23 @@
+    </div>
+</div>
+
+<button class="btn btn--tertiary print__hidden u-no-js-hide" onclick="window.print()">Print this page <span class="btn__icon btn__icon--print-link"></span></button>
```
`.print__hidden` class not needed

dwyeradam replied:
Without this *Print this page* text appears on the print preview and i'm assuming the print itself. I feel this spoils the look of a printed page.

hamishtaplin reviewed (changes requested):
Couple of minor things

hamishtaplin reviewed (changes requested):
Couple of minor things

ajmaddaford commented:
I get a `TypeError: __init__() missing 2 required positional arguments: 'ru_name' and 'ru_ref'` error when trying to continue a survey with an active session after switching to this version.

ajmaddaford commented on data/en/test_submitted_responses.json:
```diff
@@ -0,0 +1,113 @@
+{
```
It would be clearer if this file was named `test_view_submitted_response.json`

ajmaddaford commented on tests/functional/spec/features/submitted_responses.spec.js:
```diff
@@ -0,0 +1,66 @@
+const SubmissionPage = require('../../pages/features/submitted_responses/submission.page.js');
+
+
+describe('@watch Submitted Responses', function() {
```
- The `@watch` tag needs to be removed
- To be consistent with the other feature test it should be prefixed with `Feature: `

ajmaddaford commented on app/storage/storage_encryption.py:
```diff
@@ -0,0 +1,66 @@
+        if user_ik is None:
+            raise ValueError('User ik must be set')
+
+        pepper = current_app.eq['secret_store'].get_secret_by_name('EQ_SERVER_SIDE_STORAGE_ENCRYPTION_USER_PEPPER')
```
I don't think this class should access current_app, as this introduces the dependency of being in a flask app. Testing this also requires the app context whereas if pepper was passed in it could be tested in isolation.

ajmaddaford commented on app/templates/view-submission.html:
```diff
@@ -0,0 +1,23 @@
+    </div>
+</div>
+
+<button class="btn btn--tertiary print__hidden u-no-js-hide" onclick="window.print()">Print this page <span class="btn__icon btn__icon--print-link"></span></button>
```
Is this a button that looks like a link?

dwyeradam replied:
This is taken from the Design Pattern Library so we need to review this along with the opposite link looking like a button to get to the view-submission page.

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -533,7 +631,11 @@ def _get_context(full_routing_path, block, current_location, form):
+        context = {'form': form,
+                   'summary': summary_rendered_context,
+                   'variables': variables,
+                   'is_view_submission_enabled': is_view_submitted_response_enabled(g.schema.json),
```
I think we need to keep the naming consistent - is there any reason not to call the boolean passed to the template `is_view_submitted_response_enabled`?

ajmaddaford commented on app/views/questionnaire.py:
```diff
@@ -533,7 +631,11 @@ def _get_context(full_routing_path, block, current_location, form):
+                   'summary': summary_rendered_context,
+                   'variables': variables,
+                   'is_view_submission_enabled': is_view_submitted_response_enabled(g.schema.json),
+                   'is_editable': True}
```
I didn't get what this was until I viewed the template. I think something like `answers_are_editable` would be clearer.

ajmaddaford commented on docker-compose.yml:
```diff
@@ -44,10 +44,15 @@ services:
+      EQ_SUBMITTED_RESPONSES_TABLE_NAME: dev-submitted-responses
+      EQ_DYNAMODB_ENDPOINT: http://eq-docker-dynamodb:8000
+      AWS_ACCESS_KEY_ID: dummy
+      AWS_SECRET_ACCESS_KEY: dummy
```
We should discourage people from accidentally committing their AWS credentials by having a default and allowing overriding via environment variables (this is done in https://github.com/ONSdigital/eq-compose/blob/master/eq.yml#L91) e.g. `AWS_ACCESS_KEY_ID: "${AWS_ACCESS_KEY_ID:-dummy}"`

ajmaddaford commented on tests/app/data_model/test_submitted_responses.py:
```diff
@@ -0,0 +1,55 @@
+        }
+
+
+    def test_insert_item(self):
```
This needs to be renamed `test_put_item` to match the renamed method

yoganandkunche reviewed (approved):
LGTM 👍🏼

dwyeradam committed:
Allow viewing of submitted answers

- Create standalone Encrypter for QuestionnaireStorage and SessionData
- Encrypt server side session data
- Session remains active after submission and Session data used for page context
- DynamoDB to store submitted answers with ttl
- Content changes to Thank-You page
